stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_adc.h

```diff
@@ -6259,19 +6263,43 @@ __STATIC_INLINE uint32_t LL_ADC_GetOverSamplingDiscont(ADC_TypeDef *ADCx)
   */
 __STATIC_INLINE void LL_ADC_ConfigOverSamplingRatioShift(ADC_TypeDef *ADCx, uint32_t Ratio, uint32_t Shift)
 {
-  MODIFY_REG(ADCx->CFGR2, (ADC_CFGR2_OVSS | ADC_CFGR2_OVSR), (Shift | (((Ratio - 1UL) << ADC_CFGR2_OVSR_Pos))));
+#if defined(ADC_VER_V5_V90)
+    if (ADCx == ADC3)
+    {
+        /* ADC channels ratio shift oversampling */
+        MODIFY_REG(ADCx->CFGR2, (ADC_CFGR2_OVSS | ADC3_CFGR2_OVSR), (Shift | ((Ratio >> ADC3_CFGR2_OVSR_Pos))));
+    }
+    else
+#endif /* ADC_VER_V5_V90 */
+    {
+        /* ADC channels ratio shift oversampling */
+        MODIFY_REG(ADCx->CFGR2, (ADC_CFGR2_OVSS | ADC_CFGR2_OVSR), (Shift | (((Ratio - 1UL) << ADC_CFGR2_OVSR_Pos))));
+    }
 }
 
 /**
   * @brief  Get ADC oversampling ratio
   *        (impacting both ADC groups regular and injected)
+  * @note   Caution: Oversampling Ratio is dependent to ADC instance and IP version:
+  *         For STM32H72x/3x: ADC3 has 3-bit OVSR, and 9-bit OVSR for ADC1/ADC2
+  *         For Rest of STM32H7xxx All ADCs have 9-bit OVSR.
   * @rmtoll CFGR2    OVSR           LL_ADC_GetOverSamplingRatio
   * @param  ADCx ADC instance
   * @retval Ratio This parameter can be in the from 1 to 1024.
+  *               or 2, 4, 8, 16, 32, 64, 128, 256 for ADC3.
 */
 __STATIC_INLINE uint32_t LL_ADC_GetOverSamplingRatio(ADC_TypeDef *ADCx)
 {
-  return (((uint32_t)(READ_BIT(ADCx->CFGR2, ADC_CFGR2_OVSR)) + (1UL << ADC_CFGR2_OVSR_Pos)) >> ADC_CFGR2_OVSR_Pos);
+#if defined(ADC_VER_V5_V90)
+    if (ADCx == ADC3)
+    {
+        return (uint32_t)(2 << (READ_BIT(ADCx->CFGR2, ADC3_CFGR2_OVSR) >> ADC3_CFGR2_OVSR_Pos));
```

Comment:
See comment above about how value stored in register is not what will be returned by this function.